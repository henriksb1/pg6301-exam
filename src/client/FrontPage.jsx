import React from "react";

export function FrontPage() {
  return (
    <div>
      <h1>Welcome 👋🏻</h1>
    </div>
  );
}