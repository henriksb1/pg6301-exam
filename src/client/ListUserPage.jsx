import * as React from "react";
import { ErrorView } from "./ErrorView";
import { LoadingView } from "./LoadingView";
import { useLoading } from "./useLoading";
import { Link } from "react-router-dom";

export function ListUserPage({ userApi }) {
  const { data: user, error, loading, reload } = useLoading(
    async () => await userApi.listUsers()
  );

  if (error) {
    return <ErrorView error={error} reload={reload} />;
  }

  if (loading || !user) {
    return <LoadingView />;
  }

  return (
    <div>
      <h1>Edit users</h1>
      {user.map(({ id, firstName, lastName, email }) => (
        <li key={id}>
          <Link to={`/user/${id}/edit`}>
            {firstName + " " + lastName + ` (${email})`}
          </Link>
        </li>
      ))}
    </div>
  );
}
