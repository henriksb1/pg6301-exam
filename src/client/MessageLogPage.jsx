import React from "react";
import {useLoading} from "./useLoading";
import {ErrorView} from "./ErrorView";
import {LoadingView} from "./LoadingView";
import {fetchJson} from "./http";

export function MessageLogPage({messageApi}) {
    const { data: user, error, loading, reload } = useLoading(
        async () => await messageApi.listMessages()
    );

    const { data: userData, error: errorData, loading: loadingData } = useLoading(() => fetchJson("/api/profile"));

    if (error || errorData) {
        return <ErrorView error={(error ? error : errorData)} reload={reload} />;
    }

    if (loading || !user || loadingData) {
        return <LoadingView />;
    }

    const { username } = userData;
    const findSentMessages = user.filter(e => e.username === username)
    console.log(findSentMessages)
    const findReceivedMessages = user.filter(e => e.receiver === username)

    return (
        <div>
            <h1>Message log</h1>
            <h2>Sent messages</h2>
            {findSentMessages.map(({ id, username, message }) => (
                <div key={id}>
                    <p>{"Sender: " + username}</p>
                    <p>{"Message: " + message}</p>
                    <br/>
                </div>
            ))}
            <h2>Received Messages </h2>
            {findReceivedMessages.map(({ id, username, message }) => (
                <div key={id}>
                    <p>{"Sender: " + username}</p>
                    <p>{"Message: " + message}</p>
                    <br/>
                </div>
            ))}
        </div>
    )
}