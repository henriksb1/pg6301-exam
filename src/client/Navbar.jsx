import { Link } from "react-router-dom";
import * as React from "react";

export function Navbar() {
  return (
    <div className="navbar-container">
        <li><Link className="navbar-link" to={"/"}>
            Front Page
        </Link></li>
        <li><Link className="navbar-link" to={"/message"}>
            Send Message
        </Link></li>
        <li><Link className="navbar-link" to={"/log"}>
            Message Log
        </Link></li>
        <li><Link className="navbar-link" to={"/profile"}>
            Profile
        </Link></li>
        <li><Link className="navbar-link" to={"/login"}>
            Login
        </Link></li>
    </div>
  );
}
