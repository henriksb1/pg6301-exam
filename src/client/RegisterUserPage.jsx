import * as React from "react";
import { useState } from "react";
import { InputField } from "./InputField";

export function RegisterUserPage({ userApi }) {
  const [firstName, setFirstName] = useState("");
  const [lastName, setLastName] = useState("");
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");

  async function submit(e) {
    e.preventDefault();
    await userApi.createUsers({ firstName, lastName, email, password });
  }

  return (
    <form onSubmit={submit}>
      <h1>Create new user</h1>
      <InputField
        label={"First Name"}
        value={firstName}
        onValueChange={setFirstName}
      />
      <InputField
        label={"Last Name"}
        value={lastName}
        onValueChange={setLastName}
      />
      <InputField
        label={"Email Address"}
        value={email}
        onValueChange={setEmail}
      />
      <InputField
        label={"Password"}
        value={password}
        onValueChange={setPassword}
      />
      <button>Submit</button>
    </form>
  );
}
