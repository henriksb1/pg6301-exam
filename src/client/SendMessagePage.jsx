import React, {useState} from "react";
import {InputField} from "./InputField";
import {useLoading} from "./useLoading";
import {fetchJson} from "./http";
import {ErrorView} from "./ErrorView";
import {LoadingView} from "./LoadingView";

export function SendMessagePage({userApi, messageApi}) {
    //const [receiver, setReceiver] = useState("");
    const [message, setMessage] = useState("");

    const { loading, error, data } = useLoading(() => fetchJson("/api/profile"));

    if (error) {
        return <ErrorView error={error} />;
    }
    if (loading || !data) {
        return <LoadingView />;
    }

    const { username } = data;

    async function submit(e) {
        e.preventDefault();
        await messageApi.createMessage({ username, message });
    }

    return (
        <form onSubmit={submit}>
            <h1>Send new message</h1>
            <InputField
                label={"Message"}
                value={message}
                onValueChange={setMessage}
            />
            <button>Submit</button>
        </form>
    );
}