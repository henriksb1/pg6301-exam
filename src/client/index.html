<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <link rel="stylesheet" href="style.css">
    <title>Exam</title>
  </head>
  <body>
    <div id="app"></div>
  </body>
  <script src="index.jsx"></script>
</html>
