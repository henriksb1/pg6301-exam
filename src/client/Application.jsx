import { BrowserRouter } from "react-router-dom";
import { Route, Switch } from "react-router";
import { ProfilePage } from "./pages/ProfilePage";
import { LoginPage } from "./pages/LoginPage";
import { FrontPage } from "./pages/FrontPage";
import React from "react";
import { fetchJson, postJson } from "./lib/http";
import { Navbar } from "./components/Navbar";
import { SendMessagePage } from "./pages/SendMessagePage";
import { MessageLogPage } from "./pages/MessageLogPage";

export function Application() {
  const userApi = {
    listUsers: async () => await fetchJson("/api/user"),
    createUsers: async ({ firstName, lastName, email, password }) => {
      return postJson("/api/user", {
        firstName,
        lastName,
        email,
        password,
      });
    },
  };
  const messageApi = {
    listMessages: async () => await fetchJson("/api/message"),
    createMessage: async ({ username, receiver, message }) => {
      return postJson("/api/message", {
        username,
        receiver,
        message,
      });
    },
  };

  return (
    <BrowserRouter>
      <header>
        <Navbar />
      </header>
      <main>
        <Switch>
          <Route path="/profile">
            <ProfilePage userApi={userApi} />
          </Route>
          <Route path="/message">
            <SendMessagePage userApi={userApi} messageApi={messageApi} />
          </Route>
          <Route path="/log">
            <MessageLogPage messageApi={messageApi} />
          </Route>
          <Route path="/login">
            <LoginPage />
          </Route>
          <Route exact path="/">
            <FrontPage />
          </Route>
          <Route>
            <h1>Not found</h1>
          </Route>
        </Switch>
      </main>
    </BrowserRouter>
  );
}
