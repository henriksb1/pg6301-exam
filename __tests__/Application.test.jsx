import { Application } from "../src/client/Application";
import * as ReactDOM from "react-dom";
import * as React from "react";
import { act } from "react-dom/test-utils";
import { MemoryRouter } from "react-router-dom";

describe("application", () => {
  it("can show home page", async () => {
    const container = document.createElement("div");
    ReactDOM.render(
      <MemoryRouter>
        <Application />
      </MemoryRouter>,
      container
    );
    expect(container.innerHTML).toMatchSnapshot();
    expect(container.querySelector("h1").textContent).toEqual("Welcome 👋🏻");
  });

  it("can navigate to login page", async () => {
    const container = document.createElement("div");
    await act(async () => {
      await ReactDOM.render(
        <MemoryRouter>
          <Application />
        </MemoryRouter>,
        container
      );
    });
    const loginLink = [...container.querySelectorAll("a")].find(
      (a) => a.textContent === "Login"
    );
    await act(async () => {
      await loginLink.dispatchEvent(new MouseEvent("click", { bubbles: true }));
    });
    expect(container.innerHTML).toMatchSnapshot();
    expect(container.querySelector("h1").textContent).toEqual("Please log in");
  });
});
