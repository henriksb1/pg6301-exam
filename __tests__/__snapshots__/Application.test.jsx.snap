// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`application can navigate to login page 1`] = `"<header><div class=\\"navbar-container\\"><li><a class=\\"navbar-link\\" href=\\"/\\">Front Page</a></li><li><a class=\\"navbar-link\\" href=\\"/message\\">Send Message</a></li><li><a class=\\"navbar-link\\" href=\\"/log\\">Message Log</a></li><li><a class=\\"navbar-link\\" href=\\"/profile\\">Profile</a></li><li><a class=\\"navbar-link\\" href=\\"/login\\">Login</a></li></div></header><main><div><h1>Please log in</h1><form><div><label>Email: <input type=\\"text\\" value=\\"\\"></label></div><div><label>Password: <input type=\\"password\\" value=\\"\\"></label></div><button>Log in</button></form></div></main>"`;

exports[`application can show home page 1`] = `"<header><div class=\\"navbar-container\\"><li><a class=\\"navbar-link\\" href=\\"/\\">Front Page</a></li><li><a class=\\"navbar-link\\" href=\\"/message\\">Send Message</a></li><li><a class=\\"navbar-link\\" href=\\"/log\\">Message Log</a></li><li><a class=\\"navbar-link\\" href=\\"/profile\\">Profile</a></li><li><a class=\\"navbar-link\\" href=\\"/login\\">Login</a></li></div></header><main><div><h1>Welcome 👋🏻</h1><p>Please log in to use this messaging application</p></div></main>"`;
